kong/tools/utils.lua

```diff
@@ -1295,6 +1298,71 @@ do
 ]],
   }
 
+  local ngx_log = ngx.log
+  local ERR     = ngx.ERR
+  local custom_error_templates = setmetatable({}, {
+    __index = function(self, format)
+      local template_path = kong.configuration["error_template_" .. format]
+      if not template_path then
+        rawset(self, format, false)
+        return false
+      end
+
+      local template, err
+      if pl_path.exists(template_path) then
+        template, err = pl_file.read(template_path)
+      else
+        err = "file not found"
+      end
+
+      if template then
+        rawset(self, format, template)
+        return template
+      end
+
+      ngx_log(ERR, fmt("failed reading the custom %s error template: %s", format, err))
+      rawset(self, format, false)
+      return false
+    end
+  })
+
+
+  get_response_type = function(accept_header)
+    local content_type = MIME_TYPES[CONTENT_TYPE_DEFAULT]
+    if type(accept_header) == "table" then
+      accept_header = join(",", accept_header)
+    end
+
+    if accept_header ~= nil then
+      local accept_values = split(accept_header, ",")
+      local max_quality = 0
+      for _, value in ipairs(accept_values) do
```

Comment:
I would use one regular expression to parse each accept value, after stripping any white space from the beginning and the end.  According to the syntax of the accept header defined in [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.2), the following regular expression should be usable to parse the type, subtype and q-value out of any accept header:
```
([a-z0-9-!#$&^_+.]+|\*)\/([a-z0-9-!#$&^_+.]+|\*)(?:;q=(1(?:\.0{0,3}|)|0(?:\.\d{0,3}|))|)
```
I am certain that some would object to the expression as being pretty complex, but a manually written parser would probably be a lot of code and not be easier to review.

This regular expression parser would also be more correct in that it'd parse quality values `0` and `1`, which would not be parsed by yours.

I used https://regexr.com/ to develop this expression and found it pretty nifty and helpful.

Reply:
While you're at it, you could also implement the wildcard for `*` to match any type or subtype that we have content for.

Reply:
we might need to tune the regex a little to match q-values even when a `level` property is defined, e.g. `text/html;level=1` --> `([a-z0-9][a-z0-9-!#$&^_+.]+|\*)\/([a-z0-9][a-z0-9-!#$&^_+.]+|\*)(?:;level=\d|)(?:;q=(1(?:\.0{0,3}|)|0(?:\.\d{0,3}|))|)`. I'm analyzing different scenarios to ensure this is compatible with the code we currently have [in the response module](https://github.com/Kong/kong/blob/dbe8d947e7d6a6c12f8464fc2f08276be1cd392f/kong/pdk/response.lua#L1064-L1094) and cause no regressions.

Reply:
Please note this section from RFC7231:

> Each media-range might be followed by zero or more applicable media
> type parameters (e.g., charset), an optional "q" parameter for
> indicating a relative weight ([Section 5.3.1](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1)), and then zero or more
> extension parameters.  The "q" parameter is necessary if any
> extensions (accept-ext) are present, since it acts as a separator
> between the two parameter sets.

This means that an `Accept` header with `text/html;level=2;q=0.4` would not be conforming to the spec.  In that case, the regular expression would not see the `q` value, which I think would be acceptable.

Reply:
after discussing this offline we agreed to adapt the regex pattern to support said parameters.
I updated the code with the new pattern, with a very small variation where I used a single capturing group to get the media type instead of splitting type/subtype. I did it because `utils.get_mime_type` expects the media_type to be passed as a single string so there should be no need to capture type and subtype separately. The wildcard for subtype matching is already implemented in `get_mime_type` based on that input.